### PR TITLE
lib/model: More connection closing

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
Not sure wtf happened here. Did this never work sensibly?

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
The current user experience:

image

File: lib/model/model.go [hunk 1]
```diff
@@ -2426,6 +2429,7 @@ func (m *model) CommitConfiguration(from, to config.Configuration) bool {
 				l.Infoln("Adding folder", cfg.Description())
 				m.newFolder(cfg)
 			}
+			m.closeConns(cfg.DeviceIDs(), errFolderShared)
```
Comment (calmh) [inline lib/model/model.go:2432]: newFolder calls closeConns, and a new paused folder shouldn't require closing

File: lib/model/model.go [hunk 2]
```diff
@@ -2434,6 +2438,7 @@ func (m *model) CommitConfiguration(from, to config.Configuration) bool {
 		if !ok {
 			// The folder was removed.
 			m.removeFolder(fromCfg)
+			m.closeConns(fromCfg.DeviceIDs(), errFolderRemoved)
```
Comment (calmh) [inline lib/model/model.go:2441]: removeFolder calls stopFolder which calls closeConns

File: lib/model/model.go [hunk 3]
```diff
@@ -2445,13 +2450,19 @@ func (m *model) CommitConfiguration(from, to config.Configuration) bool {
...
+				m.closeConns(diff, errFolderShared)
+			}
 		}
 
 		// Emit the folder pause/resume event
 		if fromCfg.Paused != toCfg.Paused {
 			eventType := events.FolderResumed
 			if toCfg.Paused {
 				eventType = events.FolderPaused
+				m.closeConns(toCfg.DeviceIDs(), ErrFolderPaused)
```
Comment (calmh) [inline lib/model/model.go:2463]: The folder is restarted when paused/resumed, which has closeConns calls

File: lib/model/model.go [hunk 4]
```diff
@@ -2445,13 +2450,19 @@ func (m *model) CommitConfiguration(from, to config.Configuration) bool {
 		// Check if anything differs that requires a restart.
 		if !reflect.DeepEqual(fromCfg.RequiresRestartOnly(), toCfg.RequiresRestartOnly()) {
 			m.restartFolder(fromCfg, toCfg)
+			if diff := difference(fromCfg.DeviceIDs(), toCfg.DeviceIDs()); len(diff) > 0 {
```
Comment (calmh) [inline lib/model/model.go:2453]: restartFolder calls closeConns

File: lib/model/model.go (hunk 1 above)
Comment (AudriusButkevicius) [reply lib/model/model.go:2432]: It would in theory. For consistency we should send cc with paused=true

**Comment by AudriusButkevicius** [comment]
I'll follow this up with some tests.

**Comment by AudriusButkevicius** [comment]
I've added some tests and a fix for the one test case that is broken.

File: lib/model/model.go [hunk 5]
```diff
@@ -496,6 +493,8 @@ func (m *model) restartFolder(from, to config.FolderConfiguration) {
 	}
 
 	m.stopFolder(from, fmt.Errorf("%v folder %v", errMsg, to.Description()))
+	// Need to send CC change to both from and to devices.
+	m.stopFolder(to, fmt.Errorf("%v folder %v", errMsg, to.Description()))
```
Comment (imsodin) [inline lib/model/model.go:497]:
```suggestion
	err := fmt.Errorf("%v folder %v", errMsg, to.Description()))
	m.stopFolder(from, err)
	// Need to send CC change to both from and to devices.
	m.closeConns(to.DeviceIDs, err)
```

**Comment by AudriusButkevicius** [comment]
Build servers seem to be broken, can't resolve github.com

**Pull request merged**